### PR TITLE
Major refactor of helper functions

### PR DESCRIPTION
**Pull request by chrisvanrun**
This PR refactors existing helper functions (i.e. `Client.update_archive_item`) and adds some new ones that make sense.

These are the helper functions now:

- `update_archive_item`
- `add_cases_to_archive` (new)
- `update_display_set` (new)
- `add_cases_to_reader_study`
- `run_external_job`
- `upload_cases` (untouched)

## Proto models

The core of this PR is the 'gcapi/models_proto.py'.

This PR introduces proto-models. These are models of things that are yet to be. Their main function is to separate the creation/saving logic from the validation logic. This allows for quick unit tests to cover corner cases in validation, while simultaneously allowing to validate all values before anything hits the platform.

These 'preflight' validations were, partially, already in place for `add_cases_to_reader_study` but are now ubiquitous for all helper functions.

While refactoring I ran into several issues concerning the adaptation of models, which I fixed as I went. With deadline pressure and random fixes, flaky tests, et cetera this is now a messy and large change set. Sorry!

Below are some descriptions of additional fixes that are included in the PR.

---

### Both `str`, and [`str`] are allowed
Sourcing a CIV from one file no longer requires a one-lengthed list, the helper functions automagically handle both. This also applies to `pathlib.Path`.

---

### `.details` API now accepts an `api_url` keyword argument
Since we are casting to models it made more sense to be able to call the `.details` with api_url than to manually figure out which `gcapi.models` one is to use.

This replaces:

```Python
 gcapi.models.HyperLinkedImage(**client(url=api_url))
 ```
 
 With:
 ```Python
 client.images.details(api_url=api_url)
 ```
 ---

 ### `.create()`/`.update()`/`.partial_update()` -> models

 API calls that support `.create(..)`, `.update(..)`, and `.partial_update(..)` now return `gcapi.models` instead of `dicts`.

---

### Using existing images

Related to the actual pitch:
- https://github.com/DIAGNijmegen/rse-roadmap/issues/305

One can add an existing image as follows (pulled from a test):

```Python
 # First create an image component interface
added_display_sets = client.add_cases_to_reader_study(
    reader_study="reader-study",
    display_sets=[
        {
            "ct-image": [TESTDATA / "image10x10x101.mha"],
        }
    ],
)

display_set_pk = added_display_sets[0]

# ... wait for import to finish

display_set = c.reader_studies.display_sets.detail(pk=display_set_pk)

# Re-use both as a direct image and a CIV
display = display_set.values[0]
image = client.images.detail(api_url=display.image)

added_display_sets = await client.add_cases_to_reader_study(
    reader_study="reader-study",
    display_sets=[
        {
            "ct-image": image,
        },
        {
            "ct-image": civ,
        },
    ],
)
```

---
### Some typing stuff
Closes: #163

- #163
---

### Usage of `JSON` for `"File"` super_kind interfaces.
Closes: #153

- https://github.com/DIAGNijmegen/rse-gcapi/issues/153

It is now possible to provide:
- a file path to a super-kind:`"Value"` interface
- a `dict` (or other JSON-like python object) to a super-kind:`"File"` interface

Note that for the `"File"` one, this **cannot** be done for an interface with kind `"String"`. It would be far too easy to have a non-existing file path as content for a String interface. All other interfaces should error out on the platform and run into schemas there.

### CODE REVIEW

File: tests/integration_tests.py [hunk 1]
```diff
@@ -389,36 +390,17 @@ def test_add_and_update_file_to_archive_item(local_grand_challenge):
...
-        files=[Path(__file__).parent / "testdata" / "image10x10x101.mha"],
+        files=[TESTDATA / "image10x10x101.mha"],
     )
 
     # retrieve existing archive item pk
     items = get_archive_items(c, archive.pk, len(old_items_list))
 
     old_civ_count = len(items[-1].values)
 
-    with pytest.raises(ValueError) as e:
```
Comment (chrisvanrun) [inline tests/integration_tests.py:400]: Covered by unitetsts

File: tests/integration_tests.py [hunk 2]
```diff
@@ -522,116 +501,150 @@ def get_updated_archive_item_detail():
...
-                        [20, 88, 0.5],
-                        [83, 88, 0.5],
-                        [83, 175, 0.5],
-                        [20, 175, 0.5],
-                    ],
-                    "version": {"major": 1, "minor": 0},
-                },
-                "predictions-csv-file": [
-                    Path(__file__).parent / "testdata" / "test.csv"
+CIV_SET_PARAMS = (
```
Comment (chrisvanrun) [inline tests/integration_tests.py:588]: Defining this 4x seemed a bit verbose.

File: tests/integration_tests.py [hunk 3]
```diff
@@ -683,89 +696,142 @@ def check_file(interface_value, expected_name):
...
+            },
+        ],
     )
+    assert len(added_display_sets) == 2
+    for ds_pk in added_display_sets:
+        ds = c.reader_studies.display_sets.detail(pk=ds_pk)
+        assert len(ds.values) == 1
 
 
-def test_add_cases_to_reader_study_multiple_files(local_grand_challenge):
```
Comment (chrisvanrun) [inline tests/integration_tests.py:757]: Covered by unittests

**Comment by chrisvanrun** [comment]
Fingers crossed... locally everything ran butter smooth. It's the CI restraints that show the race conditions.

**Comment by chrisvanrun** [comment]
Warnings:
```
/home/runner/work/rse-gcapi/rse-gcapi/gcapi/client.py:510: DeprecationWarning: Using ["api_url"] for getting attributes is deprecated and will be removed in the next release. Suggestion: Replace ["api_url"] with .api_url
      uploads=[u["api_url"] for u in uploads], **kwargs
  tests/integration_tests.py::test_upload_cases_to_archive[files1-None]
  tests/integration_tests.py::test_upload_cases_to_archive[files0-generic-overlay]
  tests/integration_tests.py::test_upload_cases_to_archive[files2-generic-overlay]
  tests/integration_tests.py::test_upload_cases_to_archive[files3-None]
    /home/runner/work/rse-gcapi/rse-gcapi/tests/integration_tests.py:150: DeprecationWarning: Using ["pk"] for getting attributes is deprecated and will be removed in the next release. Suggestion: Replace ["pk"] with .pk
      us = get_upload_session(c, us["pk"])
  tests/integration_tests.py::test_upload_cases_to_archive[files1-None]
  tests/integration_tests.py::test_upload_cases_to_archive[files0-generic-overlay]
  tests/integration_tests.py::test_upload_cases_to_archive[files2-generic-overlay]
  tests/integration_tests.py::test_upload_cases_to_archive[files3-None]
    /home/runner/work/rse-gcapi/rse-gcapi/tests/integration_tests.py:178: DeprecationWarning: Using ["files"] for getting attributes is deprecated and will be removed in the next release. Suggestion: Replace ["files"] with .files
      response = c(url=image["files"][0]["file"], follow_redirects=True)
  tests/integration_tests.py::test_upload_cases_to_archive[files1-None]
  tests/integration_tests.py::test_upload_cases_to_archive[files0-generic-overlay]
  tests/integration_tests.py::test_upload_cases_to_archive[files2-generic-overlay]
  tests/integration_tests.py::test_upload_cases_to_archive[files3-None]
    /home/runner/work/rse-gcapi/rse-gcapi/tests/integration_tests.py:178: DeprecationWarning: Using ["file"] for getting attributes is deprecated and will be removed in the next release. Suggestion: Replace ["file"] with .file
      response = c(url=image["files"][0]["file"], follow_redirects=True)
  tests/integration_tests.py::test_download_cases[files0]
    /home/runner/work/rse-gcapi/rse-gcapi/tests/integration_tests.py:283: DeprecationWarning: Using ["pk"] for getting attributes is deprecated and will be removed in the next release. Suggestion: Replace ["pk"] with .pk
      us = get_upload_session(c, us["pk"])
  tests/integration_tests.py::test_create_job_with_upload[test-algorithm-evaluation-image-1-generic-medical-image-files0]
    /home/runner/work/rse-gcapi/rse-gcapi/tests/integration_tests.py:339: DeprecationWarning: Using ["status"] for getting attributes is deprecated and will be removed in the next release. Suggestion: Replace ["status"] with .status
      assert job["status"] == "Queued"
  tests/integration_tests.py::test_create_job_with_upload[test-algorithm-evaluation-image-1-generic-medical-image-files0]
    /home/runner/work/rse-gcapi/rse-gcapi/tests/integration_tests.py:340: DeprecationWarning: Using ["inputs"] for getting attributes is deprecated and will be removed in the next release. Suggestion: Replace ["inputs"] with .inputs
      assert len(job["inputs"]) == 1
  tests/integration_tests.py::test_create_job_with_upload[test-algorithm-evaluation-image-1-generic-medical-image-files0]
    /home/runner/work/rse-gcapi/rse-gcapi/tests/integration_tests.py:341: DeprecationWarning: Using ["pk"] for getting attributes is deprecated and will be removed in the next release. Suggestion: Replace ["pk"] with .pk
      job = c.algorithm_jobs.detail(job["pk"])
  tests/integration_tests.py::test_upload_cases_to_archive_item_with_new_interface
    /home/runner/work/rse-gcapi/rse-gcapi/tests/integration_tests.py:258: DeprecationWarning: Using ["pk"] for getting attributes is deprecated and will be removed in the next release. Suggestion: Replace ["pk"] with .pk
      us = get_upload_session(c, us["pk"])
  tests/integration_tests.py::test_upload_cases_to_archive_item_with_existing_interface
    /home/runner/work/rse-gcapi/rse-gcapi/tests/integration_tests.py:227: DeprecationWarning: Using ["pk"] for getting attributes is deprecated and will be removed in the next release. Suggestion: Replace ["pk"] with .pk
      us = get_upload_session(c, us["pk"])
  tests/async_integration_tests.py::test_download_cases[files0]
    /home/runner/work/rse-gcapi/rse-gcapi/tests/async_integration_tests.py:328: DeprecationWarning: Using ["pk"] for getting attributes is deprecated and will be removed in the next release. Suggestion: Replace ["pk"] with .pk
      us = await get_upload_session(c, us["pk"])
  tests/async_integration_tests.py::test_upload_cases_to_archive[files1-None]
  tests/async_integration_tests.py::test_upload_cases_to_archive[files2-generic-overlay]
  tests/async_integration_tests.py::test_upload_cases_to_archive[files3-None]
  tests/async_integration_tests.py::test_upload_cases_to_archive[files0-generic-overlay]
    /home/runner/work/rse-gcapi/rse-gcapi/tests/async_integration_tests.py:170: DeprecationWarning: Using ["pk"] for getting attributes is deprecated and will be removed in the next release. Suggestion: Replace ["pk"] with .pk
      us = await get_upload_session(c, us["pk"])
  tests/async_integration_tests.py::test_upload_cases_to_archive_item_with_existing_interface
    /home/runner/work/rse-gcapi/rse-gcapi/tests/async_integration_tests.py:258: DeprecationWarning: Using ["pk"] for getting attributes is deprecated and will be removed in the next release. Suggestion: Replace ["pk"] with .pk
      us = await get_upload_session(c, us["pk"])
  tests/async_integration_tests.py::test_create_job_with_upload[test-algorithm-evaluation-image-1-generic-medical-image-files0]
    /home/runner/work/rse-gcapi/rse-gcapi/tests/async_integration_tests.py:381: DeprecationWarning: Using ["status"] for getting attributes is deprecated and will be removed in the next release. Suggestion: Replace ["status"] with .status
      assert job["status"] == "Queued"
  tests/async_integration_tests.py::test_create_job_with_upload[test-algorithm-evaluation-image-1-generic-medical-image-files0]
    /home/runner/work/rse-gcapi/rse-gcapi/tests/async_integration_tests.py:382: DeprecationWarning: Using ["inputs"] for getting attributes is deprecated and will be removed in the next release. Suggestion: Replace ["inputs"] with .inputs
      assert len(job["inputs"]) == 1
  tests/async_integration_tests.py::test_create_job_with_upload[test-algorithm-evaluation-image-1-generic-medical-image-files0]
    /home/runner/work/rse-gcapi/rse-gcapi/tests/async_integration_tests.py:384: DeprecationWarning: Using ["pk"] for getting attributes is deprecated and will be removed in the next release. Suggestion: Replace ["pk"] with .pk
      job = await c.algorithm_jobs.detail(job["pk"])
  tests/async_integration_tests.py::test_upload_cases_to_archive_item_with_new_interface
    /home/runner/work/rse-gcapi/rse-gcapi/tests/async_integration_tests.py:301: DeprecationWarning: Using ["pk"] for getting attributes is deprecated and will be removed in the next release. Suggestion: Replace ["pk"] with .pk
 ```

**Comment by chrisvanrun** [comment]
How can it be that something that is hyperlinked referenced by another object from the API (e.g. via `https://gc.localhost/api/v1/cases/images/8e3e90d5-7622-4574-84fe-82bbfadcd2b7/`) gives a `404`?

In a race condition... the only thing I can readily think of is that it is a permission problem... do those get set later? To be investigated later.

**Comment by chrisvanrun** [comment]
Ok, possible fix: not recursing in tests when `ObjectNotFound` is thrown (instead of a `HTTPStatusError`)

**Comment by chrisvanrun** [comment]
Converting this to a draft and going to cut it up in smaller piece for reviewing.

**Comment by chrisvanrun** [comment]
Guh, this PR is quite interwoven. I think I'll approach this roughly as follows:

-pick out some small fixes that are VERY easy to review. The problem is I don't always remember what they fixed =P.
-extract the core system of the `ProtoCIV`, without hooking that into something. Just as an implementation.
- hook up the core system to existing helper functions.
- create new helper functions.
